File: router/memory.py
"""Unified memory retrieval (implementation.md §2, step 1 of routing policy).

Combines:
  - Graft (Trail Brain): codebase structural & project memory, via local CLI
  - Mem0: general long-term semantic memory, via platform REST API
    (active only when MEM0_API_KEY is set; otherwise skipped gracefully)

The merged context is prepended to the Jev `state` for routing decisions.
"""
from __future__ import annotations

import json
import os
import subprocess
import urllib.request
from dataclasses import dataclass, field


@dataclass
class GraftRetriever:
    """Project/code memory via the local graft CLI (`graft ask --source`)."""

    project_root: str = "."
    max_chars: int = 1500

    def retrieve(self, query: str) -> str:
        try:
            out = subprocess.run(
                ["graft", "ask", query[:400], "--source"],
                cwd=self.project_root, capture_output=True, text=True, timeout=30,
            )
        except (subprocess.TimeoutExpired, FileNotFoundError) as e:
            return f"[graft unavailable: {e}]"
        text = out.stdout.strip()
        return text[: self.max_chars] if text else "[graft: no context returned]"


@dataclass
class Mem0Retriever:
    """Long-term semantic memory via the Mem0 platform API.

    Disabled unless MEM0_API_KEY is set — the router logs a note and
    continues with Graft-only context (degraded, not failed).
    """

    top_k: int = 8
    min_score: float = 0.65  # §10 memory.mem0 config
    api_key: str = field(default_factory=lambda: os.environ.get("MEM0_API_KEY", ""))
    user_id: str = "default"

    @property
    def available(self) -> bool:
        return bool(self.api_key)

    def retrieve(self, query: str) -> str:
        if not self.available:
            return "[mem0: not configured (MEM0_API_KEY unset); graft-only context]"
        req = urllib.request.Request(
            "https://api.mem0.ai/v1/memories/search/",
            data=json.dumps({
                "query": query[:1000],
                "user_id": self.user_id,
                "filters": {"AND": [{"user_id": self.user_id}]},
                "top_k": self.top_k,
                "min_score": self.min_score,
            }).encode(),
            headers={"Authorization": f"Token {self.api_key}",
                     "Content-Type": "application/json"},
        )
        try:
            with urllib.request.urlopen(req, timeout=15) as resp:
                data = json.loads(resp.read())
        except Exception as e:  # noqa: BLE001
            return f"[mem0: search failed: {e}]"
        hits = [m.get("memory", "") for m in data.get("results", [])]
        return " | ".join(hits) if hits else "[mem0: no relevant memories]"


def retrieve_all(query: str, project_root: str = ".") -> dict:
    """Run both retrievers and return structured context for the router state."""
    graft = GraftRetriever(project_root=project_root).retrieve(query)
    mem0 = Mem0Retriever().retrieve(query)
    return {
        "graft": graft,
        "mem0": mem0,
        "mem0_active": Mem0Retriever().available,
    }


def format_context(ctx: dict) -> str:
    """Render retrieved memory into the state block fed to Jev."""
    return f"Project context (Graft):\n{ctx['graft']}\n\nLong-term memory (Mem0):\n{ctx['mem0']}"

File: router/memory.ts
/**
 * Unified memory retrieval (implementation.md §2, step 1 of routing policy).
 *
 * TypeScript port of memory.py. Combines:
 *   - Graft (Trail Brain): codebase structural & project memory, via local CLI
 *   - Mem0: general long-term semantic memory, via platform REST API
 *     (active only when MEM0_API_KEY is set; otherwise skipped gracefully)
 *
 * The merged context is prepended to the Jev `state` for routing decisions.
 */
import { spawnSync } from "node:child_process";
import type { JsonTable } from "../evals/harness/jev_policy.ts";

/** Project/code memory via the local graft CLI (`graft ask --source`). */
export function graftRetrieve(query: string, projectRoot = ".", maxChars = 1500): string {
  let out;
  try {
    out = spawnSync("graft", [query.slice(0, 400), "--source"], {
      cwd: projectRoot,
      encoding: "utf8",
      timeout: 30_000,
    });
  } catch (error) {
    return `[graft unavailable: ${error}]`;
  }
  if (out.error) return `[graft unavailable: ${out.error}]`;
  const text = (out.stdout ?? "").trim();
  return text.length > 0 ? text.slice(0, maxChars) : "[graft: no context returned]";
}

export interface Mem0RetrieverOptions {
  topK?: number;
  minScore?: number;
  apiKey?: string;
  userId?: string;
}

/** Long-term semantic memory via the Mem0 platform API.
 *
 * Disabled unless MEM0_API_KEY is set — the router logs a note and
 * continues with Graft-only context (degraded, not failed). */
export class Mem0Retriever {
  readonly topK: number;
  readonly minScore: number; // §10 memory.mem0 config
  readonly apiKey: string;
  readonly userId: string;

  constructor(options: Mem0RetrieverOptions = {}) {
    this.topK = options.topK ?? 8;
    this.minScore = options.minScore ?? 0.65;
    this.apiKey = options.apiKey ?? process.env["MEM0_API_KEY"] ?? "";
    this.userId = options.userId ?? "default";
  }

  get available(): boolean {
    return Boolean(this.apiKey);
  }

  async retrieve(query: string): Promise<string> {
    if (!this.available) {
      return "[mem0: not configured (MEM0_API_KEY unset); graft-only context]";
    }
    try {
      const resp = await fetch("https://api.mem0.ai/v1/memories/search/", {
        method: "POST",
        headers: { Authorization: `Token ${this.apiKey}`, "Content-Type": "application/json" },
        body: JSON.stringify({
          query: query.slice(0, 1000),
          user_id: this.userId,
          filters: { AND: [{ user_id: this.userId }] },
          top_k: this.topK,
          min_score: this.minScore,
        }),
        signal: AbortSignal.timeout(15_000),
      });
      const data = (await resp.json()) as { results?: Array<{ memory?: string }> };
      const hits = (data["results"] ?? []).map((m) => m["memory"] ?? "");
      return hits.length > 0 ? hits.join(" | ") : "[mem0: no relevant memories]";
    } catch (error) {
      return `[mem0: search failed: ${error}]`;
    }
  }
}

/** Run both retrievers and return structured context for the router state. */
export async function retrieveAll(query: string, projectRoot = "."): Promise<JsonTable> {
  const graft = graftRetrieve(query, projectRoot);
  const mem0 = new Mem0Retriever();
  const mem0Out = await mem0.retrieve(query);
  return {
    graft,
    mem0: mem0Out,
    mem0_active: mem0.available,
  };
}

/** Render retrieved memory into the state block fed to Jev. */
export function formatContext(ctx: JsonTable): string {
  return `Project context (Graft):\n${ctx["graft"]}\n\nLong-term memory (Mem0):\n${ctx["mem0"]}`;
}
